### PR TITLE
Avoid possible recursions

### PR DESCRIPTION
**Pull request by T-256**
# Summary
Discussed in https://github.com/encode/httpx/discussions/2772

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [ ] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.

### CODE REVIEW

File: httpx/_client.py [hunk 1]
```diff
@@ -1719,6 +1719,8 @@ async def _send_single_request(self, request: Request) -> Response:
             response = await transport.handle_async_request(request)
 
         assert isinstance(response.stream, AsyncByteStream)
+        # We are not accept BoundAsyncStream here to avoid recursion
```
Comment (T-256) [inline httpx/_client.py:1722]:
```suggestion
        # Avoid recursion of `response.stream`
```

**Comment by tomchristie** [comment]
Sorry I don't understand this pull request?...

File: httpx/_client.py [hunk 2]
```diff
@@ -1719,6 +1721,9 @@ async def _send_single_request(self, request: Request) -> Response:
             response = await transport.handle_async_request(request)
 
         assert isinstance(response.stream, AsyncByteStream)
+        # Avoid recursion of `response.stream`
+        assert not isinstance(response.stream, BoundAsyncStream)
+
         response.request = request
         response.stream = BoundAsyncStream(
             response.stream, response=response, timer=timer
```
[the remark this answers is not in the document]
Comment (tomchristie) [reply httpx/_client.py:1729]: > Here, if `response.stream` has already bounded via BoundAsyncStream we don't need bound it again

So it's not clear to me... is this a bug in `pytest-httpx` and `respx`?
If so, is this due to some unintended private API usage?
Or perhaps something that only reproduces under `asyncio`, but is okay with `trio` and the sync version?

> You can reproduce this recursion as said https://github.com/encode/httpx/discussions/2772

Let's always try to break these down into simpler cases.
Here's an example that'll trigger the assertion you've introduced in this pull request...

```python
import httpx
import asyncio
import pytest
import pytest_httpx


@pytest.mark.asyncio
async def test_something(httpx_mock):
    httpx_mock.add_response(json={"abc": "def"})
    async with httpx.AsyncClient() as client:
        tasks = [client.get("https://example.com/") for idx in range(3)]
        responses = await asyncio.gather(*tasks)
    print(responses)
```

Requires `httpx`, `pytest` and `pytest-httpx` to be installed.
Run with `pytest example.py`.

[the remark this answers is not in the document]
Comment (T-256) [reply httpx/_client.py:1729]: > So it's not clear to me... is this a bug in `pytest-httpx` and `respx`?

Only in `pytest-httpx`. As I [tested](https://github.com/encode/httpx/discussions/2772#discussioncomment-6460472),  the test passed successfully in `respx`.

Actually it's bug with `pytest-httpx`, but we can specify that it should not use recursion at here.

[the remark this answers is not in the document]
Comment (tomchristie) [reply httpx/_client.py:1729]: > Actually it's bug with `pytest-httpx`

Are you able to isolate how/where?

> we can specify that it should not use recursion at here.

Can you describe what "it should not use recursion" means *specifically*?
I can see the assertion, which makes sense to me, but we've not yet pointed at a "don't do *this*" example.

[the remark this answers is not in the document]
Comment (T-256) [reply httpx/_client.py:1729]: > Are you able to isolate how/where?
> 
No, I didn't debug inside pytest-httpx.

> Can you describe what "it should not use recursion" means _specifically_?

when `response.stream` already is `BoundAsyncStream` that means it's already bounded, so should not done again.

> we've not yet pointed at a "don't do _this_" example.

Correct, we should find the wrong example at first. (needs debugging pytest-httpx)

[the remark this answers is not in the document]
Comment (tomchristie) [reply httpx/_client.py:1729]: Okay...

A more simple reproduction is:

```python
import httpx
import pytest
import pytest_httpx


@pytest.mark.asyncio
async def test_something(httpx_mock):
    httpx_mock.add_response(json={"abc": "def"})
    with httpx.Client() as client:
        client.get("https://example.com/")
        client.get("https://example.com/")
        client.get("https://example.com/")
```

The issue is essentially that a *single* response instance is being returned by the transport, and being reused.

[the remark this answers is not in the document]
Comment (T-256) [reply httpx/_client.py:1729]: > The issue is essentially that a _single_ response instance is being returned by the `httpx_mock` transport, and being reused.

Yes, it uses single `response` object:
https://github.com/Colin-b/pytest_httpx/blob/d6a58906015ef76e5e86c101b8f65851b237f409/pytest_httpx/_httpx_mock.py#L127-L137
Then httpx changing `response.stream` incrementaly.

There are few solutions:
1. Don't allow `transport.handle_request` to return `Response` object that `stream` attribute be `BoundSyncStream`. (Current PR)
2. Auto extract main stream if `Response.stream` is `BoundSyncStream`:
```py
stream = response.stream
if isinstance(stream, BoundSyncStream):
    stream = stream.stream
response.stream = BoundAsyncStream(
    stream, response=response, timer=timer
)
```

[the remark this answers is not in the document]
Comment (T-256) [reply httpx/_client.py:1729]: I can guess second option will destroy concurrent streams and we can ignore it.

Do you think documentation needs modify at [mock-transports](https://www.python-httpx.org/advanced/#mock-transports)?

**Comment by T-256** [comment]
I think new version of `pytest-httpx` fixed https://github.com/encode/httpx/discussions/2772. do we still need this assertation included in the PR?
I'm closing this now. for future if any bigger problem has occurred, we can reference to here.

**Pull request closed without merging**